library/cc/engine.cc

```diff
@@ -6,23 +6,26 @@
 namespace Envoy {
 namespace Platform {
 
-Engine::Engine(envoy_engine_t engine, const std::string& configuration, LogLevel log_level)
-    : engine_(engine), terminated_(false) {
-  run_engine(this->engine_, configuration.c_str(), logLevelToString(log_level).c_str());
-
-  this->stream_client_ = std::make_shared<StreamClient>(this->engine_);
-  this->pulse_client_ = std::make_shared<PulseClient>();
+Engine::Engine(envoy_engine_t engine) : engine_(engine), terminated_(false) {}
+
+// we lazily construct the stream and pulse clients
+// because they either require or will require a weak ptr
+// which can't be provided from inside of the constructor
+// because of how std::enable_shared_from_this works
+StreamClientSharedPtr Engine::streamClient() {
+  if (!this->stream_client_) {
+    this->stream_client_ = std::make_shared<StreamClient>(this->weak_from_this());
```

Comment:
I think StreamClient should probably take a shared reference to the Engine rather than a weak one. We made stats weak so that all the stats objects that might be distributed around a program wouldn't all be retaining the engine... but then again, it's up for debate whether they should be, as well.

Reply:
(Also FWIW, it _does_ effectively take a shared ref in the other platforms.)

Reply:
This is a weak ptr to avoid a cycle (`Engine` -> `StreamClient` -> `Engine`), but I don't think this cycle needs to exist anymore. It's just a byproduct of the original design of this PR where `Stream`s kept `Engine`s alive.

I can address this comment along with the earlier comment at the same time, I think, by just removing the `EngineSharedPtr` references throughout `StreamClient` and `StreamPrototype` and just move them back to `envoy_engine_t`